# eBay Kleinanzeigen

The aim of this project is to clean the data and analyze the included used car listings.

The data dictionary provided with data is as follows:

- `dateCrawled`: When this ad was first crawled. All field-values are taken from this date.
- `name`: Name of the car.
- `seller`: Whether the seller is private or a dealer.
- `offerType`: The type of listing
- `price`: The price on the ad to sell the car.
- `abtest`: Whether the listing is included in an A/B test.
- `vehicleType`: The vehicle Type.
- `yearOfRegistration`: The year in which the car was first registered.
- `gearbox`: The transmission type.
- `powerPS`: The power of the car in PS.
- `model`: The car model name.
- `kilometer`: How many kilometers the car has driven.
- `monthOfRegistration`: The month in which the car was first registered.
- `fuelType`: What type of fuel the car uses.
- `brand`: The brand of the car.
- `notRepairedDamage`: If the car has a damage which is not yet repaired.
- `dateCreated`: The date on which the eBay listing was created.
- `nrOfPictures`: The number of pictures in the ad.
- `postalCode`: The postal code for the location of the vehicle.
- `lastSeenOnline`: When the crawler saw this ad last online.

In [31]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv',encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## Dataframe brief description

Keypoints on the dataframe:
- the `name` column contains a free string. Some of the items first word is the car brand, but that's not a standard.
- the column `price` has a few items with the value 0, we need to check this for errors. The same happens with the `powerPS` and `montheOfRegistration` columns
- `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage` have null values. This must be taken into account.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

These changes were made by two reasons:
- to comply with Python's snake case standard;
- to make some of the names more readable

Let's move on to do some basic data exploration.

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

In [9]:
autos['registration_month'].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

## Dataframe short description

- columns that have mostly one value (candidates to be dropped): `seller` and `offer_type` have two unique values, with one of them accounting for 49,999 of the results; `num_photos` has 0 for all its rows. We will drop these three columns.
- columns that need more investigation: `registration_year` has numbers representing years in the future (e.g.: 4100 and 4800) or in a distant past (e.g.: 1001 and 1111); `power_ps` has numbers representing odd horsepower (above 2000); `registration_month` has the value 0; `vehicle_type`, `gearbox`, `model`, `fuel_type` and `unrepaired_damage` have null values.
- columns with numeric data stored as text: `price` and `odometer`.

We will start by changing `price` and `odometer` to numeric values.

In [10]:
autos.drop(['num_photos','seller','offer_type'],axis=1,inplace=True)

In [11]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)

In [12]:
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)

In [13]:
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)

Further look into `price` and `odometer_km`:

In [14]:
autos['price'].nunique()

2357

In [15]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
autos['price'].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

By looking at the top 15 values in `price` we can see that the ones above $350,000 are unrealistic, so we will drop them.

In [17]:
autos = autos[autos['price'].between(0,350000)]

In [18]:
autos['odometer_km'].nunique()

13

In [19]:
autos['odometer_km'].describe()

count     49986.000000
mean     125736.506222
std       40038.133399
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head(15)

150000    32416
125000     5169
100000     2168
90000      1757
80000      1436
70000      1230
60000      1164
50000      1025
40000       818
30000       789
20000       784
10000       264
5000        966
Name: odometer_km, dtype: int64

No problems have been verified for `odometer_km`, unless numbers are rounded (maybe a tendency from the sellers or a site's imposition).

Further look into columns with dates:

In [21]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [22]:
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

The database crawled between 2016-03-05 and 2016-04-07.

In [23]:
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033229
2016-03-10    0.031869
2016-03-11    0.032789
2016-03-12    0.036610
2016-03-13    0.016925
2016-03-14    0.035230
2016-03-15    0.033749
2016-03-16    0.030008
2016-03-17    0.031189
2016-03-18    0.013724
2016-03-19    0.033849
2016-03-20    0.037871
2016-03-21 

The database contains ads created between 2015-06-11 and 2016-04-07

In [24]:
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64

In [25]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As mentioned before, `registration_year` has numbers representing years in the future (e.g.: 4100 and 4800) or in a distant past (e.g.: 1001 and 1111).

In [26]:
autos[autos['registration_year'].between(1900,2016)].shape[0] / autos.shape[0]

0.9605889649101749

Cars between 1900 (when cares were roughly invented) and 2016 (when the database was crawled) account for 96% of the database, so we will drop the rest.

In [27]:
autos = autos[autos['registration_year'].between(1900,2016)]

In [79]:
autos['brand'].value_counts() > 1000

volkswagen         True
bmw                True
opel               True
mercedes_benz      True
audi               True
ford               True
renault            True
peugeot            True
fiat               True
seat              False
skoda             False
mazda             False
nissan            False
smart             False
citroen           False
toyota            False
sonstige_autos    False
hyundai           False
volvo             False
mini              False
mitsubishi        False
honda             False
kia               False
alfa_romeo        False
porsche           False
suzuki            False
chevrolet         False
chrysler          False
dacia             False
daihatsu          False
jeep              False
subaru            False
land_rover        False
saab              False
jaguar            False
trabant           False
daewoo            False
rover             False
lancia            False
lada              False
Name: brand, dtype: bool

In [113]:
mean_prices = {}
total_cars = {}

for brand in autos['brand'].value_counts().index:
    brand_df = autos[autos['brand'] == brand]
    mean_price = brand_df['price'].mean()
    total_car = (autos['brand'] == brand).sum()
    mean_prices[brand] = round(mean_price)
    total_cars[brand] = total_car

In [114]:
average_prices = pd.Series(mean_prices)
tot_cars = pd.Series(total_cars)

In [118]:
#df = pd.DataFrame(average_prices,columns=['mean_price'])
#df['total_cars'] = pd.Series(total_cars)
df = pd.DataFrame(tot_cars,columns=['total_cars'])
df['mean_price'] = pd.Series(average_prices)
df.sort_values('total_cars',inplace=True,ascending=False)
df

total_cars  mean_price
volkswagen           10185        5231
bmw                   5283        8103
opel                  5194        2877
mercedes_benz         4579        8485
audi                  4149        9094
ford                  3350        3652
renault               2274        2395
peugeot               1418        3039
fiat                  1242        2712
seat                   873        4296
skoda                  770        6335
mazda                  727        4011
nissan                 725        4665
smart                  668        3543
citroen                668        3700
toyota                 599        5115
sonstige_autos         523       10805
hyundai                473        5309
volvo                  444        4757
mini                   415       10460
mitsubishi             391        3334
honda                  377        3988
kia                    341        5789
alfa_romeo             318        3985
porsche                293       44553
suzuki                 284        3996
chevrolet              274        6489
chrysler               176        3229
dacia                  123        5916
daihatsu               123        1556
jeep                   108       11435
subaru                 105        3765
land_rover              98       19108
saab                    77        3212
jaguar                  76       11176
trabant                 75        1552
daewoo                  72        1020
rover                   65        1528
lancia                  52        3246
lada                    29        2503

We can observe a large gap in the mean price of the top brand (Porsche) and the second one (Land Rover).

We can also verify that the most listed car brand (Volkswagen) accounts for almost the double number of cars of the brand in second position (BMW). The top 5 brands are all German, and that makes sense :)

From now on, we will focus our analysis on the top 6 brands in terms of listed ads.

In [123]:
top_list = list(df.head(6).index)
top_list

['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford']

In [131]:
mean_prices = {}
mean_mileage = {}

for brand in top_list:
    brand_df = autos[autos['brand'] == brand]
    mean_price = brand_df['price'].mean()
    mean_mil = brand_df['odometer_km'].mean()
    mean_prices[brand] = round(mean_price)
    mean_mileage[brand] = round(mean_mil)

In [140]:
df = pd.DataFrame(pd.Series(mean_prices),columns=['average_price'])
df['average_mileage'] = pd.Series(mean_mileage)
df.sort_values('average_price',ascending=False)

average_price  average_mileage
audi                    9094           129288
mercedes_benz           8485           130856
bmw                     8103           132431
volkswagen              5231           128724
ford                    3652           124069
opel                    2877           129223

While the top 3 brands are close to each other in terms of price, we can see a large gap for the other two, being Opel the least expensive. Volkswagen stands somehow in the middle.

Average mileage turns out to be irrelevant as they are very close to each other.

#### Finding the most common pairs Brand/Model:

In [144]:
autos['brand_model'] = autos['brand'] + ' ' + autos['model']

In [146]:
autos['brand_model'].value_counts().head()

volkswagen golf      3815
bmw 3er              2688
volkswagen polo      1677
opel corsa           1645
volkswagen passat    1388
Name: brand_model, dtype: int64

#### Finding the difference in average price between cars with and without damage:

In [149]:
autos['unrepaired_damage'].value_counts().index

Index(['nein', 'ja'], dtype='object')

In [152]:
damage_dict = {}

for damage in autos['unrepaired_damage'].value_counts().index:
    damage_df = autos[autos['unrepaired_damage'] == damage]
    average_price = damage_df['price'].mean()
    damage_dict[damage] = round(average_price)
    
damage_dict

{'nein': 7077, 'ja': 2127}